profiler/src/main/java/com/splunk/opentelemetry/profiler/snapshot/PeriodicStackTraceSampler.java

```diff
@@ -0,0 +1,234 @@
+      }
+    }
+
+    private void sample(Collection<SamplingContext> contexts) {
+      if (contexts.isEmpty()) {
+        return;
+      }
+
+      Map<Long, SamplingContext> threadContexts =
+          contexts.stream().collect(Collectors.toMap(c -> c.thread.getId(), context -> context));
+      long currentSampleTime = System.nanoTime();
+      try {
+        ThreadInfo[] threadInfos = collector.getThreadInfo(threadContexts.keySet());
+        List<StackTrace> stackTraces =
+            toStackTraces(threadInfos, threadContexts, currentSampleTime);
+        staging.get().stage(stackTraces);
+      } catch (Exception e) {
+        logger.info("Unexpected error during callstack sampling");
+      }
+    }
+
+    private List<StackTrace> toStackTraces(
+        ThreadInfo[] threadInfos, Map<Long, SamplingContext> contexts, long currentSampleTime) {
+      List<StackTrace> stackTraces = new ArrayList<>(threadInfos.length);
+      for (ThreadInfo threadInfo : threadInfos) {
+        SamplingContext context = contexts.get(threadInfo.getThreadId());
+        // When the context is locked and on demand sample is being taken. No need to report
+        // both so skip the on-demand sample.
+        if (context.lock.tryLock()) {
+          try {
+            SpanContext spanContext = retrieveActiveSpan(context.thread);
+            toStackTrace(threadInfo, context, spanContext.getSpanId(), currentSampleTime)
+                .ifPresent(stackTraces::add);
+          } finally {
+            context.lock.unlock();
+          }
+        }
+      }
+      return stackTraces;
+    }
+
+    private Optional<StackTrace> toStackTrace(
+        ThreadInfo threadInfo, SamplingContext context, String spanId, long currentSampleTime) {
+      // If multiple threads have managed to take a sample for the same context
+      // one of the sampling periods may be a negative value. If this happens a
+      // previous sample fully encompasses this sample and so this sample can
+      // be safely dropped.
+      Duration samplingPeriod = Duration.ofNanos(currentSampleTime - context.sampleTime);
+      if (samplingPeriod.isNegative()) {
+        return Optional.empty();
+      }
+      context.updateSampleTime(currentSampleTime);
+      return Optional.of(
+          StackTrace.from(
+              Instant.now(),
+              samplingPeriod,
+              threadInfo,
+              context.traceId,
+              spanId,
+              Thread.currentThread().getId()));
+    }
+
+    /** It's possible the active span will have changed since the sample was taken */
+    private SpanContext retrieveActiveSpan(Thread thread) {
+      return spanTracker.get().getActiveSpan(thread).orElse(SpanContext.getInvalid());
+    }
+  }
+
+  private static class SamplingContext {
+    private final Lock lock = new ReentrantLock();
+    private final Thread thread;
+    private final String traceId;
+    private final String spanId;
+    private volatile long sampleTime;
```

Comment:
as far as I can tell this is now always updated when holding a lock, so doesn't need to be volatile